### DISCUSSION TITLE
Welcome to Pode.Web Discussions!

### DISCUSSION THREAD

**Discussion by Badgerati** [Badgerati/Pode.Web#344] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas, or tools you've built using Pode.Web.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

**Comment by Conrad-awaze** [comment]
Is there a way of passing the $WebEvent.Data output from one ScriptBlock to another in a different container?

**Reply by Badgerati** [reply]
Hi @Conrad-awaze,

There isn't a direct way of doing this, though it depends on what you're trying to achieve. If it's to use $WebEvent.Data to update elements in another container then this is possible (depending on the element).

Have you an example of what you're trying to achieve?

Thanks!

**Reply by Conrad-awaze** [reply]
Thanks for the quick reply.

I'm trying to take  values from 2 New-PodeWebSelect options in a WebForm and use them to generate and table in a WebTable in a WebCard. 

Each Server has different accounts so the table will populate depending which server and database is passed in. Here is the script as an example. and I've attached it as well with a screenshot of the Web Server.

```PowerShell
New-PodeWebForm -Name 'Servers' -Content @(
            New-PodeWebSelect -Name 'Server' -Options 'Server1', 'Server2' 
            New-PodeWebSelect -Name 'Database' -Options 'Database 1', 'Database 2' 
        ) -ScriptBlock {
            $ServerName = $WebEvent.Data['Server'] 
            $DatabaseName = $WebEvent.Data['Database']    
        }
   
        New-PodeWebCard -Content @(
            New-PodeWebTable -Name 'Accounts' -ArgumentList $ServerName, $DatabaseName -DataColumn 'UserID' -ScriptBlock {
                param($ServerName,$DatabaseName)

            }
        )
```  

The $ServerName and $DatabaseName variables are the ones I need to generate the WebTable. If there is another way of doing this then I'm open to suggestions.   

Thanks in advance.  

[DBA_DashboardV5.txt](https://github.com/Badgerati/Pode.Web/files/11517432/DBA_DashboardV5.txt)
![Web Server](https://github.com/Badgerati/Pode.Web/assets/109225751/69551c49-7d20-4110-b502-9ee47e6744f5)

**Comment by thisisricky** [comment]
Has this been resolved? The only way I found to make it work was by using global variables. 

When I tried to query `$($WebEvent.Data)` from the table Card, to stop a service for example, the Data key is empty.

Example Code
```powershell
Add-PodeWebPage -Name 'Services' -Icon 'Cogs' -Group "Computer MGMT" -ScriptBlock {
        New-PodeWebContainer -Content @(
            New-PodeWebForm -Name "Host FQDN" -Id "ComputerNameText" -SubmitText "GO" -ShowReset -ResetText "Clear" -Content @(
                        New-PodeWebTextbox -Name "ComputerName" -DisplayName "ComputerName" -Id "ComputerName"
                    ) -ScriptBlock {
                        [array]$Output = @()
                        if ([string]::IsNullOrEmpty($WebEvent.Data.ComputerName))
                        {
                            Show-PodeWebToast "The ComputerName field cannot be empty" -Title "Error"
                            return
                        }
                        else {
                            if(Test-Connection $($WebEvent.Data.ComputerName) -quiet -count 1)
                            {
                                Show-PodeWebToast -Message "Connecting to $($($WebEvent.Data.ComputerName))"
                                try {
                                    $services = Invoke-command -Computername $($WebEvent.Data.ComputerName) -ScriptBlock{ Get-Service} -Credential $($MyCred) -ErrorAction Stop
                                    
                                }
                                Catch {
                                    Show-PodeWebToast -Message "Failed to connect to $($($WebEvent.Data.ComputerName))"
                                    return
                                }
                            }
                            else {
                                Show-PodeWebToast -Message "Host $($WebEvent.Data.ComputerName) is unreachable" -Title "Error"
                                return
                            }
                        }
                        foreach($srv in $services)
                        {
                            $Output += [ordered]@{
                                ComputerName = $($WebEvent.Data.ComputerName)
                                Name      = $srv.Name
                                Status    = "$($srvStatus)"
                                StartType = "$($srv.StartType)"
                                Actions   = @(
                                    New-PodeWebButton -Name 'Stop' -Icon 'Stop-Circle' -Colour Red -Size Small -Outline -ScriptBlock {
                                        Invoke-command -Computername $($WebEvent.Data.ComputerName) -ScriptBlock{
                                            param($ServiceName) 
                                            Stop-Service -Name $ServiceName
                                        } -ArgumentList $($WebEvent.Data.Value) -Credential $($MyCred)
                                        $WebEvent | out-default
                                        New-PodeWebAlert -Type Info -Value "$($WebEvent.Data.Value) stopped"
                                        Sync-PodeWebTable -Id $ElementData.Parent.ID
                                    }
                                    New-PodeWebButton -Name 'Start' -Icon 'Play-Circle' -Colour Green -Size Small -Outline -ScriptBlock {
                                        Invoke-command -Computername $($WebEvent.Data.ComputerName) -ScriptBlock{
                                            param($ServiceName) 
                                            Start-Service -Name $ServiceName
                                        } -ArgumentList $($WebEvent.Data.Value) -Credential $($MyCred)
                                        Show-PodeWebToast -Message "$($WebEvent.Data.Value) started"
                                        Sync-PodeWebTable -Id $ElementData.Parent.ID
                                    }
                                )
                            } 
                        } 
                        $Output | Out-PodeWebTable
                    }    
        )
    }
```
The example code above display the services as expected. Please see the image below
![image](https://github.com/Badgerati/Pode.Web/assets/80350677/069f0dba-ec26-4faf-b8f8-2016d04fbc22)

But when trying to interact with the buttons, the `$($WebEvent.Data)` is blank.

![image](https://github.com/Badgerati/Pode.Web/assets/80350677/c16cf09e-babd-4d0a-98f6-c91af022f98f)

Any help would be greatly appreciated.

**Reply by Badgerati** [reply]
Hi @thisisricky,

The issue will be with `Out-PodeWebTable`, in the current version this is simply for rendering adhoc data that's not for interacting with. What you'll want to do is create a new empty table with `New-PodeWebTable` and assign the DataColumn parameter (ie: ServiceName). In your Form's scriptblock you can then use `Update-PodeWebTable` to populate the table's data.

One issue here is that you can currently only have 1 DataColumn assigned. One way around this could be if you're using authentication o your pages, or just have sessions enabled generally. You could store the ComputerName in the user's session, and then pass the ServiceName tothe button via the table's DataColumn and retrieve the ComputerName from the user's session.

**Reply by thisisricky** [reply]
@Badgerati , Thanks for the quick reply! I'll give that a go. 

Pode is awesome by the way.

**Comment by krishna19944** [comment]
Hi @Badgerati,

Good Day.

Thank you for Pode & Pode.Web. Its an awesome tool. 

I was exploring and trying to find a layout with auto refresh. Will this feature going to add in upcoming versions? 
In general I am looking for a feature which auto reload number of elements in Pode Web Page.

Appreciates your response on this.

**Reply by Badgerati** [reply]
Hi @krishna19944,

Is there any particular elements or scenarios you're after to trigger an auto-reload? Some elements do support auto-reloading, such as tables / tiles, while others will have to be done using a combination of Update functions and Pode.Web's Timer elements.

Thanks.

**Reply by krishna19944** [reply]
Dear @Badgerati,

Thank you for your response. 

I am looking to dynamically adding new web pages or elements(like textbox, select etc) without restarting a Pode Server.

**Comment by kjebu** [comment]
Hi @Badgerati,
Is there a way to dynamically enable/disable a textbox? I'm trying to enable a text box based on which radio button the user clicks.

Thank you in advance for any insight into this .

**Reply by krishna19944** [reply]
Dear @kjebu,

You may explore Show-PodeWebComponent & Hide-PodeWebComponent with onclick event of Form/button to achieve this.

**Reply by kjebu** [reply]
Thanks a lot @krishna19944!  This was exactly what I was looking for :-)

Kjetil
